packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -851,16 +884,34 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
     private attachListener() {
         this.setMaxListeners(Number.MAX_SAFE_INTEGER);
         this.dataStoreContext.once("attaching", () => {
-            assert(this.bindState !== BindState.NotBound,
-                0x186 /* "Data store attaching should not occur if it is not bound" */);
+            /**
+             * back-compat 0.58.2000 - Ideally, attachGraph() should have already been called making the data store
+             * locally visible. However, before visibility state was added, this may not have been the case and data
+             * store can move to "attaching" state in 2 scenarios:
+             * 1) Before attachGraph() is called - When a data store is created and bound in an attached container.
+             * 2) After attachGraph() is called - When a detached container is attached.
+             *
+             * The basic idea is that all local object should become locally visible before they are globally visible.
+             */
+            this.attachGraph();
+
             this._attachState = AttachState.Attaching;
+
+            assert(this.visibilityState === VisibilityState.LocallyVisible,
+                `Data store should not attach if is not locally visible - ${this.visibilityState}`);
```

Comment:
Data store should only attach if it is locally visible? In my mind, if the visibilityState is globally visible also means it's locally visible.

Reply:
This should say `Data should store not become globally visible without being locally visible first`.
Do you mean someone attaches the listener after the data store is globally visible? This does happen for remote data stores. They are always globally visible. But for them, "attaching" event should never fire. If it does, its a bug and we want to fail fast.